huggingface/chat-ui
# Add initial support for web browsing

https://github.com/huggingface/chat-ui/assets/25119303/d56b4f56-099d-4718-8577-b92a3f21a55c

## Why
Since the model has a knowledge cutoff and has a tendency to hallucinate it would be great to complement the chat history with relevant results from the web.

## How
We use serpapi for performing the search query. The LLM determines which query best answers the question, and we pass this query to serpapi. The results are parsed and injected into the conversation prompt. 

The web search feature can be toggled using a switch. If the key is not defined, the feature is disabled globally.



## What's missing
* Reading the content of pages. Right now we only inject the headlines from google and the text snippet that goes with it. The answers would be a lot more relevant if it could read the actual pages.
* Some UI indication that something is going on. Inferring the search query and then executing the search takes time and right now to the user it's not obvious what is happening.
* Maybe some rate limit mechanism on search-enhanced queries, to avoid burning through the API credits too fast ?

## How to test
* Create a [serpapi](https://serpapi.com/) account
* Get you api key and paste it in `.env.local` under `SERPAPI_KEY`
* You can now ask questions with and without web search by toggling the switch in the ui.
* You can see what search query is used and what snippets get injected in the prompt in the console.
